WIP

LostKobrakai opened the pull request:
I hope you don't mind, but I already worked a bit on the things we discussed today.

LostKobrakai committed:
WIP

benwilson512 commented on lib/fable/events.ex:
```diff
@@ -95,6 +194,7 @@ defmodule Fable.Events do
   end
 
+  @doc false
   def log(config, agg) do
```
We do need docs for these functions somewhere.

LostKobrakai replied:
Yeah, I would've started by adding those as well to the module docs, maybe document the functions within `__using__`. I'm not really sure how useful it would eb to call functions directly on `Fable.Events`.

LostKobrakai replied:
What we could also do is add those functions to the behaviour. That's how ecto documents functions for `Ecto.Repo` and I guess they also don't really expect people to actually reimplement it (besides maybe for tests).

benwilson512 commented on test/events_test.exs:
```diff
@@ -0,0 +1,48 @@
+
+  describe "emit/2" do
+    test "returns a zero arity callback with an query as first argument" do
+      query = from a in Aggregate, where: a.id > 0
```
Repo.serial was designed to work with queryables mostly because I use that function within my code base for other stuff. For this particular purpose, we definitely do NOT want to support queryables, only actual aggregates. The query you have here locks the entire table, which is not ideal.

LostKobrakai replied:
Yeah, didn't really think of that. I just read the `repo.one` and thought it would just affect that one result. I could see using a queryable being useful, but I doubt there's a good way to make it work. Adding `limit: 1` sounds terrible and otherwise we cannot really ensure only one row is selected.

benwilson512 commented on lib/fable/events.ex:
```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
-        raise error
+  def emit(config, schema_or_queryable, fun, changes \\ %{}, opts)
+
+  def emit(%{repo: repo} = config, %Ecto.Query{} = queryable, fun, changes, opts)
```
Per the other comment on queryables, I can't see a reason why we should support queryables.

benwilson512 commented on lib/fable/events.ex:
```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
+
+    fn ->
+      schema = lock(schema, repo) || schema
+      rollback_on_error(repo, handle_events(config, schema, fun.(schema, repo, changes), opts))
```
as a minor style thing, I'd probably prefer to yank out the `fun.()` call and the `handle_events` call here into named variables.

benwilson512 commented on lib/fable/events.ex:
```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
+          Keyword.t()
+        ) ::
+          Ecto.Multi.t()
+  def emit(config, multi, schema_or_queryable, name, fun, opts)
```
We should pattern match on the ecto multi struct here.

benwilson512 commented on lib/fable/events.ex:
```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
+          Ecto.Multi.t()
+  def emit(config, multi, schema_or_queryable, name, fun, opts)
+      when is_function(fun, 3) do
+    Ecto.Multi.run(multi, name, fn _repo, changes ->
```
We may want to do a check that `repo` is the same as `config.repo` here.

LostKobrakai replied:
Yeah, it's probably a good idea. I also added this PR to ecto: https://github.com/elixir-ecto/ecto/pull/3088. With this it would in the future be possible to decouple the event emitting part completely from the need of having a repo configured at compile time. I guess it's still needed for the process manager part though.

benwilson512 replied:
Oh nice, love that PR. And yeah you do need it for the process manager bits, although conceivably that could be passed in as an argument to the supervisor tree entry instead of hard coded at compile time.

benwilson512 commented on lib/fable/events.ex:
```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
+    do: handle_events(config, aggregate, [event], opts)
+
+  defp handle_events(config, aggregate, events, opts) when is_list(events) do
+    reduce_while_ok(aggregate, events, fn aggregate, event ->
```
You could eliminate the previous clause by doing `reduce_while_ok(aggregate, List.wrap(events),`

benwilson512 reviewed (changes requested):
This is fantastic, thanks for taking a look at it!

LostKobrakai reviewed (commented):
https://github.com/CargoSense/fable/pull/3/files#diff-6cb6f29533068060651c7c8062fe2bcaR203
What would you think of ordering by `:prev_event_id` with nulls first? Then this wouldn't be coupled to incrementing ids.

LostKobrakai commented:
Ok, scratch that question on the ordering. If events don't have incrementing ids they also don't have incrementing last_event_ids 🙈.

LostKobrakai committed:
Hopefully finished changes to Fable.Events

- Extract router behaviour
- Add events behaviour
- Complete options documentation
- Refactor log/replay functions

benwilson512 commented:
Thank you!
